### DISCUSSION TITLE
Window selection magic

### DISCUSSION THREAD

**Discussion by JustSimplyKyle** [hyprwm/Hyprland#416] [Show and tell]
`hyprctl clients -j | jq -r ".[] | select(.workspace.id == "$(hyprctl activewindow -j | jq -r '.workspace.id')\)""| jq -r ".at,.size" | jq -s "add" | jq '_nwise(4)' | jq -r '"\(.[0]),\(.[1]) \(.[2])x\(.[3])"'| slurp`
This "simple" command will will do the exact same thing as this on sway!
`swaymsg -t get_tree | jq -r '.. | select(.pid? and .visible?) | .rect | "\(.x),\(.y) \(.width)x\(.height)"' | slurp`
I'll admit, it's quite a bit uglier but it works! you can select indiviual windows using just your mouse!

**Comment by JustSimplyKyle** [comment]
For use in wayshot, this magic command.
`wayshot -s "$(hyprctl clients -j | jq -r ".[] | select(.workspace.id == "$(hyprctl activewindow -j | jq -r '.workspace.id')\)""| jq -r ".at,.size" | jq -s "add" | jq '_nwise(4)' | jq -r '"\(.[0]),\(.[1]) \(.[2])x\(.[3])"'| slurp -f '%x %y %w %h')" --stdout | wl-copy`

**Comment by JustSimplyKyle** [comment]
There are two main difficulties making this command, first is that hyprctl doesn't seem to have a way to see whether or not a window is visable. So I have to get the id of the current activewindow(which should be *always* visable) and compare it to all the other clients.
The second one is just doing a bunch of strange json magic to make it the way that slurp accepts.(which is a bit of a hassle because hyprlands selection output is quite a bit different from slurp's input)

**Comment by JustSimplyKyle** [comment]
Full script.
```bash
#/usr/bin/env bash
if [ "$(hyprctl activewindow)" = "Invalid" ] ; then
	wayshot --stdout | wl-copy
else
	if [[ -z "$(hyprctl clients -j | jq -r ".[] | select(.workspace.id == "$(hyprctl activewindow -j | jq -r '.workspace.id')\)" and select(.size == [2560,1440])")" ]];then
		hyprctl --batch "keyword decoration:active_opacity 1;keyword decoration:inactive_opacity 1";wayshot -s "$(hyprctl clients -j | jq -r ".[] | select(.workspace.id == "$(hyprctl activewindow -j | jq -r '.workspace.id')\)""| jq -r ".at,.size" | jq -s "add" | jq '_nwise(4)' | jq -r '"\(.[0]),\(.[1]) \(.[2])x\(.[3])"'| slurp -f '%x %y %w %h')" --stdout | wl-copy;hyprctl reload
	else
		hyprctl --batch "keyword decoration:active_opacity 1;keyword decoration:inactive_opacity 1";wayshot --stdout | wl-copy
	fi
fi
```
Mainly the features added are detection for if the window is fullscreen or not
*Technically* I could use `hyprctl mointors` to grep the resolution but ehh, I'm lazy

**Comment by NotAShelf** [comment]
sick script, how'd you like adding this to the contrib repo

**Reply by JustSimplyKyle** [reply]
9 days late lol
https://github.com/hyprwm/contrib/commit/94d84beed19ba2021733203efc17cc5273b9ab0a

**Comment by EmmChriss** [comment]
For those who use multiple monitors and would like a one-liner, here's my modified command. It's basically the same that @JustSimplyKyle wrote, but gets workspace list from `monitors.activeWorkspace`
```
hyprctl clients -j | jq -r ".[] | select(.workspace.id | IN($(hyprctl -j monitors | jq 'map(.activeWorkspace.id) | join(",")' | tr -d \")))" | jq -r ".at,.size" | jq -s "add" | jq '_nwise(4)' | jq -r '"\(.[0]),\(.[1]) \(.[2])x\(.[3])"'| slurp
```

And here's the hyprland bind I use with grim and swappy
```
bindl = , Print, exec, grim -g "$(hyprctl clients -j | jq -r ".[] | select(.workspace.id | IN($(hyprctl -j monitors | jq 'map(.activeWorkspace.id) | join(",")' | tr -d \")))" | jq -r ".at,.size" | jq -s "add" | jq '_nwise(4)' | jq -r '"\(.[0]),\(.[1]) \(.[2])x\(.[3])"'| slurp)" - | swappy -f -
```

**Comment by jamespullar** [comment]
`_nwise()` no longer seems to be available as a private method in jq. Here's a reworking of the solution without needing any private methods.
```
hyprctl clients -j | jq -r ".[] | select(.workspace.id == $(hyprctl activewindow -j | jq -r '.workspace.id')) | \"\(.at[0]),\(.at[1]) \(.size[0])x\(.size[1])\""
```

Here's a full wrapping with grim, slurp, and satty:
```
grim -g "$(hyprctl clients -j | jq -r ".[] | select(.workspace.id == $(hyprctl activewindow -j | jq -r '.workspace.id')) | \"\(.at[0]),\(.at[1]) \(.size[0])x\(.size[1])\"" | slurp)" - | satty -f -
```
